server/src/main/java/org/elasticsearch/TransportVersion.java

```diff
@@ -25,38 +25,43 @@
+ * the major/minor versions of {@link Version}, but {@code TransportVersion} does not have separate major/minor version numbers. So the
+ * minimum compatible version is hard-coded as the transport version used by the highest minor release of the previous major version. {@link
+ * #MINIMUM_COMPATIBLE} should be updated appropriately whenever a major release happens.
  * <p>
- * The earliest CCS compatible version is hardcoded at {@link #MINIMUM_CCS_VERSION}, as the transport version used by the
- * previous minor release. This should be updated appropriately whenever a minor release happens.
+ * The earliest CCS compatible version is hardcoded at {@link #MINIMUM_CCS_VERSION}, as the transport version used by the previous minor
+ * release. This should be updated appropriately whenever a minor release happens.
  * <h2>Adding a new version</h2>
- * A new transport version should be added <em>every time</em> a change is made to the serialization protocol of one or more classes.
- * Each transport version should only be used in a single merged commit (apart from BwC versions copied from {@link Version}).
+ * A new transport version should be added <em>every time</em> a change is made to the serialization protocol of one or more classes. Each
+ * transport version should only be used in a single merged commit (apart from BwC versions copied from {@link Version}).
  * <p>
- * To add a new transport version, add a new constant at the bottom of the list that is one greater than the current highest version,
- * ensure it has a unique id, and update the {@link CurrentHolder#CURRENT} constant to point to the new version.
+ * To add a new transport version, add a new constant at the bottom of the list that is one greater than the current highest version, ensure
+ * it has a unique id, and update the {@link CurrentHolder#CURRENT} constant to point to the new version.
  * <h2>Reverting a transport version</h2>
- * If you revert a commit with a transport version change, you <em>must</em> ensure there is a <em>new</em> transport version
- * representing the reverted change. <em>Do not</em> let the transport version go backwards, it must <em>always</em> be incremented.
+ * If you revert a commit with a transport version change, you <em>must</em> ensure there is a <em>new</em> transport version representing
+ * the reverted change. <em>Do not</em> let the transport version go backwards, it must <em>always</em> be incremented.
+ * <h2>Scope of usefulness of {@link TransportVersion}</h2>
+ * {@link TransportVersion} is a property of the transport connection between a pair of nodes, and should not be used as an indication of
+ * the version of any single node. The {@link TransportVersion} of a connection is negotiated between the nodes via some logic that is not
+ * totally trivial, and may change in future. Any other places that might make decisions based on this version effectively have to reproduce
```

Comment:
Worth including a link to the handshaker class for posterity?

Reply:
`TransportHandshaker` is package-private so we can't link to it properly. IMO that's a feature, not a bug :)